lib/aba/batch.rb

```diff
@@ -47,56 +47,68 @@ def initialize(attrs = {}, transactions = [])
     end
 
     def to_s
-      raise RuntimeError, 'No transactions present - add one using `add_transaction`' if @transactions.empty?
+      raise RuntimeError, 'No entries present - add one using `add_transaction` or `add_return`' if entries.empty?
       raise RuntimeError, 'ABA data is invalid - check the contents of `errors`' unless valid?
 
       # Descriptive record
       output = "#{descriptive_record}\r\n"
 
       # Transactions records
-      output += @transactions.map{ |t| t[1].to_s }.join("\r\n")
+      output += entries.map { |t| t[1].to_s }.join("\r\n")
 
       # Batch control record
       output += "\r\n#{batch_control_record}"
     end
 
     def add_transaction(attrs = {})
-      if attrs.instance_of?(Aba::Transaction)
-        transaction = attrs
-      else
-        transaction = Aba::Transaction.new(attrs)
-      end
-      @transactions[@transaction_index] = transaction
-      @transaction_index += 1
-      transaction
+      add_entry(Aba::Transaction, attrs)
+    end
+
+    def add_return(attrs = {})
+      add_entry(Aba::Return, attrs)
+    end
+
+    def transactions
+      entries.select { |entry| entry.instance_of?(Aba::Transaction) }
     end
 
     def transactions_valid?
-      !has_transaction_errors?
+      !transactions.map { |t| t[1].valid? }.include?(false)
+    end
+
+    def entries_valid?
+      !has_entry_errors?
     end
 
     def valid?
-      !has_errors? && transactions_valid?
+      !has_errors? && entries_valid?
     end
 
     def errors
       # Run validations
       has_errors?
-      has_transaction_errors?
+      has_entry_errors?
 
       # Build errors
       all_errors = {}
       all_errors[:aba] = self.error_collection unless self.error_collection.empty?
-      transaction_error_collection = @transactions.each_with_index.map{ |(k, t), i| [k, t.error_collection] }.reject{ |e| e[1].nil? || e[1].empty? }.to_h
-      all_errors[:transactions] = transaction_error_collection unless transaction_error_collection.empty?
+      entry_error_collection = entries.map { |k, t| [k, t.error_collection] }.reject { |e| e[1].nil? || e[1].empty? }.to_h
+      all_errors[:entries] = entry_error_collection unless entry_error_collection.empty?
 
       all_errors unless all_errors.empty?
     end
 
     private
 
-    def has_transaction_errors?
-      @transactions.map{ |t| t[1].valid? }.include?(false)
+    def add_entry(type, attrs)
+      (attrs.instance_of?(type) ? attrs : type.new(attrs)).tap do |entry|
+        entries[@entries_index] = entry
+        @entries_index += 1
```

Comment:
It doesn't look like the `entries_index` is used anywhere.

Consider getting rid of it and simply using `<<`:

```suggestion
        entries << entry
```

Reply:
Sorry I just noticed that `entries` is a hash but it might not be a problem for `entries` to be a simple array. Haven't looked at it too deeply though.

Reply:
Agree - I don't see why we couldn't use an array (at a quick glance) which implicitly has an index, however I think for now we should leave this as it is and open a new issue to discuss/review.